Internal #576: strptime strftime infinities

hawkfish opened the pull request:
Make it so we can round trip infinite temporal values through the datetime parser and formatter.
This now differs from PG but I think they are wrong.

fixes: duckdblabs/duckdb-internal#576

hawkfish committed:
Internal #576: strptime strftime infinities

Make it so we can round trip infinite temporal values
through the datetime parser and formatter.
This now differs from PG but I think they are wrong.

fixes: duckdblabs/duckdb-internal#576

hawkfish commented:
Rebasing required deleting the old branch and PR - sorry!

Mytherin commented on src/function/scalar/strftime_format.cpp:
```diff
@@ -733,7 +731,33 @@ bool StrpTimeFormat::Parse(string_t str, ParseResult &result) const {
+	//	Check for specials
 	idx_t pos = 0;
+	result.is_special = false;
+	if (Date::TryConvertDateSpecial(data, size, pos, Date::PINF)) {
```
This seems to reduce performance of `benchmark/micro/cast/strptime.benchmark` by around 10~15%, perhaps we can be a bit more clever in when we take this branch (e.g. check if the first character is a character?)

hawkfish replied:
The benchmark is too fast for consistent results but profiling with Instruments showed a definite hot spot there.  Checking for size and the first character being alpha or minus removed the hot spot.

Mytherin replied:
Great! When I ran it I increased the number of rows generated in the benchmark to double-check that it was not just noise

Mytherin reviewed (commented):
Thanks for the PR! One comment:

hawkfish committed:
Merge branch 'feature' into strptime-infinity

hawkfish committed:
Internal #576: strptime strftime infinities

CR feedback: Precheck for specials' leading characters to avoid
multiple strcmp passes on every parsed value.

samansmink reviewed (approved):
LGTM!

Mytherin commented:
Thanks!
